patches/tModLoader/Terraria/ModLoader/ModType.cs

```diff
@@ -0,0 +1,34 @@
+﻿namespace Terraria.ModLoader
+{
+	/// <summary>
+	/// The base type for most modded things.
+	/// </summary>
+	public abstract class ModType : IModType
+	{
+		///<summary>
+		/// The mod this belongs to.
+		/// </summary>
+		public Mod Mod { get; internal set; }
+
+		/// <summary>
+		/// The internal name of this.
+		/// </summary>
+		public virtual string Name => GetType().Name;
+
+		void ILoadable.Load(Mod mod) {
+			Mod = mod;
+			Load();
+			AddInstance();
+		}
```

Comment:
This separation seems to lack justification. From what I can see in this PR, `Load` is only further overridden once in `ModDust` and the code there could easily go in `AddInstance`

I suggest removing both `ILoadable.Load` and `AddInstance` and just using the single `Load(Mod)` override be inherited and passed up and down. 

~~You should also make `Mod` `private set;`~~ so people can't forget to call `base.Load(mod)` for long. Not that many people will be making custom `ModType` subclasses

Still no great way to do 'cloning'/construction, but that's another day's problem.

Reply:
Load was left for modders to overload for any code they need since Autoload() is no longer available. If you'd rather it use a single Load, I can do that. I wanted Mod to have a private set, but the cloning stuff needs it. Perhaps we could generalize the cloning into ModType.